### PR TITLE
[SYCL][Matrix] syntax changes as preparation before moving joint matrix from experimental namespace

### PR DESCRIPTION
**Pull request by yubingex007-a11y**
As part of the effort to move joint matrix from experimental namespace to supported. A review of the API is being done as part of https://github.com/intel/llvm/pull/7964. This results in the following changes in the syntax: 
1- Add Td to joint_matrix_mad as Tc can be different from Td on the GPU, Now, we make D as an input argument to mad.
2-  Change “packed” to ext_intel_packed:
3-  Move EWOps (get_wi_data, wi_element, get_coord) to detail namespace) 4- add const to joint_matrix in store and mad
5 - add joint_matrix_copy/assignment function
6- add apply with coordination (change existing tests) 
7- change get_coord vector type from int32_t to size_t
8- delete explicitly both = and copy ctor.

### CODE REVIEW

File: sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp
Comment (dkhaldi) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:92]: We should not advise users to use this. When does this runtime error occur?

File: sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp
Comment (dkhaldi) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:133]: Use `joint_matrix_apply`

File: sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp
Comment (dkhaldi) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:254]: Minor: you dont need to specify the whole namespace here

File: sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp
Comment (yubingex007-a11y) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:92]: ok, i see. wi_data class here is in sycl::ext::oneapi::experimental::matrix namespace and it is for NV. the errmsg is for intel users who uses NV's wi_data

Comment (yubingex007-a11y) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:92]: how about 
get_wi_data is available using: ext::oneapi::detail::get_wi_data but intel users are expected to use joint_matrix_copy instead

**Comment by yubingex007-a11y** [comment]
sycl/test/matrix/matrix-bfloat16-test-coord-basicB.cpp previously failed
sycl/test-e2e/Matrix/get_coord_int8_matB.cpp  failed previously
sycl/test-e2e/Matrix/element_wise_irreg_sum_rows.cpp  can't be modified easily since:
```
             for (int i = 0; i < data.length() / (TK / 4); i++) { // 4 per row
               // i*SG_SIZE index is found based on the round robin
               // distribution we are using in the implementation
               sum_local_rows[row + global_idx * (TK / 4)] += data[i + row * 4];
             }

```

**Comment by yubingex007-a11y** [comment]
I am WIP on rebasing

**Comment by yubingex007-a11y** [comment]
will handle cuda testcase later

File: sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp
Comment (dkhaldi) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:102]: Just say: "get_wi_data is unavailable, use joint_matrix_copy instead."

File: sycl/test/matrix/query-use.cpp
Comment (dkhaldi) [inline sycl/test/matrix/query-use.cpp:1]: This is an old version of the test

**Comment by dkhaldi** [comment]
> sycl/test/matrix/matrix-bfloat16-test-coord-basicB.cpp previously failed sycl/test-e2e/Matrix/get_coord_int8_matB.cpp failed previously sycl/test-e2e/Matrix/element_wise_irreg_sum_rows.cpp can't be modified easily since:
> 
> ```
>              for (int i = 0; i < data.length() / (TK / 4); i++) { // 4 per row
>                // i*SG_SIZE index is found based on the round robin
>                // distribution we are using in the implementation
>                sum_local_rows[row + global_idx * (TK / 4)] += data[i + row * 4];
>              }
> ```

We should probably remove this test because it assumes some distribution. Also, it does the same thing as sycl/test-e2e/Matrix/get_coord_int8_matB.cpp

**Comment by dkhaldi** [comment]
I looked more carefully at https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/Matrix/element_wise_irreg_sum_rows_impl.hpp
this test should be removed as it is duplicate of get_coord_matB and assumes some coordinates.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp
Comment (yubingex007-a11y) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:43]: TODO: we should modify cuda's code since we add "const" qualifier to jm_store's jm. i added mutable here as a temp fix/
the mutable keyword is used to indicate that a particular member variable of a class is allowed to be modified even when the class object is considered const.
@JackAKirk , will you take care of cuda's changes  since "const" qualifier is added to jm_store's jm?

Comment (JackAKirk) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:43]: Is this really required? the cuda implementation of `joint_matrix_store` does not modify the `joint_matrix` argument.
Can you give some more details why you think this change is required? Thanks

Comment (yubingex007-a11y) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:43]: actually, we can't pass const join_matrix's member join_matrix_cuda(it is also const ) to joint_matrix_store_cuda:
```
void joint_matrix_store_cuda(
    joint_matrix_cuda<
        T, sycl::ext::oneapi::experimental::matrix::use::accumulator, NumRows,
        NumCols, sycl::ext::oneapi::experimental::matrix::layout::dynamic> &src,
    multi_ptr<T, Space, IsDecorated> dst, size_t stride,
    sycl::ext::oneapi::experimental::matrix::layout Layout)
```
if i add "const " to joint_matrix_store_cuda's src, there is other issues too.

Comment (JackAKirk) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:43]: I don't understand but I will check out this branch and see what errors there are.

Comment (JackAKirk) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:43]: OK I understand the issue now. The fix is simple but requires quite a lot of repetitive code changes. I think it is fine to do as you suggest and use `mutable` temporarily.

Comment (JackAKirk) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:43]: I can make a follow up PR to fix it correctly.

**Comment by bso-intel** [comment]
@yubingex007-a11y 
Is the failure unrelated to your PR?

**Comment by yubingex007-a11y** [comment]
> @yubingex007-a11y Is the failure unrelated to your PR?

seems testing infrastructure issue it is not related with the PR
lit.py: /__w/llvm/llvm/llvm/sycl/test-e2e/lit.cfg.py:429: error: Cannot detect device aspect for ext_oneapi_hip:gpu

**Comment by yubingex007-a11y** [comment]
the fail is not related.

**Review by bso-intel — APPROVED** [review]

File: sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp
Comment (YuriPlyakhin) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp]: We should never advise users to use anything from detail namespace. Detail namespace are implementation details and can change at any time. It is not part of documented API.

File: sycl/include/sycl/ext/oneapi/matrix/matrix-intel.hpp
Comment (YuriPlyakhin) [inline sycl/include/sycl/ext/oneapi/matrix/matrix-intel.hpp:557]: return statement is unnecessary, please remove.
```suggestion
```

**Review by YuriPlyakhin — CHANGES_REQUESTED** [review]

File: sycl/test-e2e/Matrix/element_wise_all_ops_half_impl.hpp
Comment (yubingex007-a11y) [reply sycl/test-e2e/Matrix/element_wise_all_ops_half_impl.hpp:44]: i saw some sycl testcases are using detail's thing:
llvm/sycl/test/regression/async_work_group_copy.cpp
could we use detail::get_wi_data in these elemwise testcase? @dkhaldi @YuriPlyakhin

File: sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp
Comment (yubingex007-a11y) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:107]: i remember at least we need to make compilation pass for host.
we provide different version of device code for nv and intel., we need to provide unified host implementation of wi_data.
i remember there is a discussion but i can't find it in https://github.com/intel/llvm/pull/7413

File: sycl/test-e2e/Matrix/element_wise_all_ops_half_impl.hpp
Comment (YuriPlyakhin) [reply sycl/test-e2e/Matrix/element_wise_all_ops_half_impl.hpp:44]: In my opinion, no, we can not use detail namespace in end-to-end tests (sycl/test-e2e). We should use only public/documented API in end-to-end tests.

File: sycl/test-e2e/Matrix/element_wise_all_ops_int8_packed_impl.hpp
Comment (YuriPlyakhin) [inline sycl/test-e2e/Matrix/element_wise_all_ops_int8_packed_impl.hpp:177]: Same as above. In my opinion, test can be reduced to one joint_matrix_apply call.

File: sycl/test-e2e/Matrix/element_wise_irreg_sum_rows_impl.hpp
Comment (YuriPlyakhin) [inline sycl/test-e2e/Matrix/element_wise_irreg_sum_rows_impl.hpp:47]: if element_wise_irreg_sum_rows.cpp was removed, I think this file also should be removed.

Comment (YuriPlyakhin) [reply sycl/test-e2e/Matrix/element_wise_irreg_sum_rows_impl.hpp:47]: Also, corresponding test for XMX8 should be removed as well. What about Legacy folder, would it be reasonable to remove it from Legacy as well for consistency and to avoid any potential future maintenance?

File: sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp
Comment (yubingex007-a11y) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:132]: __SYCL2020_DEPRECATED pop comp warning while "__attribute__(( unavailable" pop compfail.
we can't let intel users to use ext::oneapi::experimental::matrix::get_wi_data. so "__attribute__(( unavailable" is reasonable. @dkhaldi , right?

File: sycl/test-e2e/Matrix/element_wise_all_ops_half_impl.hpp
Comment (yubingex007-a11y) [reply sycl/test-e2e/Matrix/element_wise_all_ops_half_impl.hpp:44]: if so, we can delete them in sycl/test-e2e. maybe we can keep them in internal testing?

Comment (dkhaldi) [reply sycl/test-e2e/Matrix/element_wise_all_ops_half_impl.hpp:44]: In the current format, this test is testing each operation, as internally get_wi_data and each operation on it will be tested. So I don't understand the proposed change to this test by @YuriPlyakhin

File: sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp
Comment (dkhaldi) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:132]: We should change the Nvidia deprecate message as well. Deprecation means it is still supported. get_wi_data is not supported at all. We removed and replaced it with joint_matrix_apply. So a better message for all backends can be:
"`get_wi_data()` has been removed from the API and replaced with `joint_matrix_apply`"

**Comment by yubingex007-a11y** [comment]
NOTES for me:
1. use "get_wi_data() has been removed from the API and replaced with joint_matrix_apply" in all messages
2. rebase

**Review by dkhaldi — APPROVED** [review]
LGTM

**Comment by yubingex007-a11y** [comment]
@intel/llvm-gatekeepers ping?

**Comment by steffenlarsen** [comment]
@YuriPlyakhin has requested changes approval is needed.

**Comment by yubingex007-a11y** [comment]
@YuriPlyakhin could you approve the pr? Dounia has answered your comments above and if there should be small changes, we can create a new pr.

**Review by JackAKirk — APPROVED** [review]

File: sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp
Comment (mmoadeli) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:43]: On HIP it works without the need for mutable.
https://github.com/intel/llvm/pull/11485/commits/42e0c62ee83be46257f8f742ff9a97c5ef4e7a62

Comment (JackAKirk) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:43]: > On HIP it works without the need for mutable. [42e0c62](https://github.com/intel/llvm/commit/42e0c62ee83be46257f8f742ff9a97c5ef4e7a62)

cuda requires reinterpret casts between signed / unsigned builtins for load/store, this is the origin of the problem. I don't think it affects AMD since amd  doesn't have builtins for load/store

Comment (mmoadeli) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:43]: there might be some similarities, I think. Initially it worked only with mutable. A number of changes needed to make it work without mutable.

Comment (JackAKirk) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:43]: oh yeah sure. I will make follow up changes for cuda after this is merged too.

**Comment by dkhaldi** [comment]
@intel/llvm-gatekeepers, please help merge

**Comment by dkhaldi** [comment]
> > @intel/llvm-gatekeepers, please help merge
> 
> There is an input from Steffen above.

Correct, I missed that. Yury is OOO today but this can wait. 
We should just make sure this gets merged before https://github.com/intel/llvm/pull/11485 is merged

**Comment by dm-vodopyanov** [comment]
> > > @intel/llvm-gatekeepers, please help merge
> > 
> > 
> > There is an input from Steffen above.
> 
> Correct, I missed that. Yury is OOO today but this can wait. We should just make sure this gets merged before #11485 is merged

Updated the description of https://github.com/intel/llvm/pull/11485

File: sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp
Comment (YuriPlyakhin) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:122]: Was this addressed?

Comment (dkhaldi) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:122]: This will be addressed by @JackAKirk among other CUDA changes.

Comment (JackAKirk) [reply sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp:122]: Yes I will make this change as soon as this PR is merged.

**Comment by YuriPlyakhin** [comment]
@intel/llvm-gatekeepers , I approved, please, merge.

**Pull request merged**
